src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -161,115 +185,176 @@ static int gavl_intersect_cmp_func(uintptr_t ustart, uintptr_t len, gavl_tree_no
+                TREE_STACK_PUSH(tree_ptr, cur_node);
+                cur_node = cur_node->right;
+                continue;
             } else {
-                /* we cannot insert the duplicate buffer */
-                gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
-                MPL_free(node_ptr);
-                goto fn_exit;
+                break;
             }
+        } else {
+            /* node match */
+            break;
+        }
+    } while (1);
 
-            if (direction == SEARCH_LEFT)
-                cur_node->left = node_ptr;
-            else
-                cur_node->right = node_ptr;
-            node_ptr->parent = cur_node;
+    *cmp_ret_ptr = cmp_ret;
+    tree_ptr->cur_node = cur_node;
+    return cur_node;
+}
+
+/* if avl tree is possibly unbalanced, gavl_tree_rebalance should be called to rebalance
+ * it. In unbalanced avl tree, the height difference between left and right child is at
+ * most 2; gavl_tree_rebalance takes it as a premise in order to rebalance tree correcly */
+static void gavl_tree_rebalance(gavl_tree_s * tree_ptr)
+{
+    gavl_tree_node_s *cur_node = tree_ptr->cur_node;
 
-          stack_recovery:
+    if (cur_node) {
+        do {
             gavl_update_node_info(cur_node);
 
             int lheight = cur_node->left == NULL ? 0 : cur_node->left->height;
             int rheight = cur_node->right == NULL ? 0 : cur_node->right->height;
             if (lheight - rheight > 1) {
+                /* find imbalance: left child is 2 level higher than right child */
                 gavl_tree_node_s *lnode = cur_node->left;
                 int llheight = lnode->left == NULL ? 0 : lnode->left->height;
+                /* if left child's left child causes this imbalance, we need to perform right rotation;
+                 * else we need to perform left-right rotation for rebalance */
                 if (llheight + 1 == lheight)
                     gavl_right_rotation(cur_node, lnode);
                 else
                     gavl_left_right_rotation(cur_node, lnode);
             } else if (rheight - lheight > 1) {
+                /* find imbalance: right child is 2 level higher than left child */
                 gavl_tree_node_s *rnode = cur_node->right;
                 int rlheight = rnode->left == NULL ? 0 : rnode->left->height;
+                /* if right child's left child causes this imbalance, we need to perform right-left
+                 * rotation; else we need to perform left rotation for rebalance */
                 if (rlheight + 1 == rheight)
                     gavl_right_left_rotation(cur_node, rnode);
                 else
                     gavl_left_rotation(cur_node, rnode);
             }
 
-            if (!STACK_EMPTY(node_stack)) {
-                STACK_POP(node_stack, cur_node);
-                goto stack_recovery;
+            /* restore search traverse trace */
+            if (!TREE_STACK_IS_EMPTY(tree_ptr)) {
```

Comment:
I don't understand what "restore search traverse trace " means. Do you mean "rebalance the previous node in the traversing trace"?

Reply:
yes, it means we have to rebalance all the nodes in the stack (or say rebalance all the nodes along traverse trace)

Reply:
BTW, I addressed your comment by reversing my comment sequence instead of changing the if else code. I didn't miss it.
> The following comment is not addressed.

> You might also want to change if (inorder_node->parent != dnode) { to if (inorder_node->parent == dnode) { and swap the two branches to match the sequence of comment.

> Besides, I think the comments can be also compressed. How about: